src/mlpack/methods/sparse_autoencoder/maximal_inputs.cpp

```diff
@@ -0,0 +1,83 @@
+#include "maximal_inputs.hpp"
+
+namespace mlpack {
+namespace nn {
+
+namespace {
+
+void VisualizeHiddenUnit(size_t rows, size_t cols,
+                         int squareRows,
+                         int offset,
+                         arma::mat const &input,
+                         arma::mat &output)
+{
+  int k = 0;
+  for(int i = 0; i != cols; ++i)
+  {
+    for(int j = 0; j != rows; ++j)
+    {
+      if(k >= input.n_cols)
+      {
+        continue;
+      }
+      arma::mat reshapeMat(squareRows, squareRows);
+      arma::mat const weights = input.row(k);
+      std::copy(std::begin(weights),
+                std::end(weights),
+                std::begin(reshapeMat));
+      double const max = arma::abs(input.row(k)).max();
+      if(max != 0.0)
+      {
+        reshapeMat /= max;
+      }
+      output.submat(i*(offset), j*(offset),
+                    i*(offset) + squareRows - 1,
+                    j*(offset) + squareRows - 1) = reshapeMat;
+      ++k;
+    }
+  }
+}
+
+}
+
+void MaximalInputs(arma::mat const &parameters, arma::mat &output)
+{
+  //take the encoder part of the paramters
+  arma::mat paramTemp = parameters.submat(0, 0, (parameters.n_rows-1)/2-1, parameters.n_cols-2);
```

Comment:
If you did transpose the parameters, here would be the place to do it (just add `.t()` to the end of the expression).  I don't think it's reasonable to avoid a copy here, since you're subtracting the mean from `paramTemp`.